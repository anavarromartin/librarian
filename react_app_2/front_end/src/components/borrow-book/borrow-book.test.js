import {mount} from "enzyme/build";
import React from "react";
import BorrowBook from "./borrow-book";

const OFFICE_ID = 1
const OFFICE_NAME = 'dallas'

const mountComponent = ({
                            setBackLocation = () => {},
                            setHeaderVisibility = () => {},
                            getAvailableBooks = () => {},
                            borrowBook = () => {}
                        } = {}) => (
    mount(<BorrowBook
        history={{push: () => {}}}
        office={{id: OFFICE_ID, name: OFFICE_NAME}}
        setBackLocation={setBackLocation}
        setHeaderVisibility={setHeaderVisibility}
        getAvailableBooks={getAvailableBooks}
        borrowBook={borrowBook}
    />)
)

const clickOnFirstSearchResult = component => {
    component.find(".search-results__row").simulate('click')
}

const typeOnSearchInput = (text, component) => {
    component.find(".labels-input__input").simulate('change', {target: {value: text}})
}

const typeOnNameInput = (text, component) => {
    component.find(".borrow__name input").simulate('change', {target: {value: text}})
}

const typeOnEmailInput = (text, component) => {
    component.find(".borrow__email input").simulate('change', {target: {value: text}})
}

const clickOnBorrowButton = component => {
    component.find(".borrow__button").simulate('click')
}

describe('<BorrowBook />', () => {

    it('renders according to snapshot', () => {
        expect(mountComponent()).toMatchSnapshot()
    })

    describe('borrow book flow', () => {
        let component
        let borrowedBookId, borrowerName, borrowerEmail, officeIdUsedToSearch, termUsedToSearch
        beforeEach(async (done) => {
            const availableBooksPromise = Promise.resolve(
                [
                    {
                        id: 1,
                        book_name: 'a book'
                    }
                ]
            )
            component = mountComponent({
                getAvailableBooks: (actualOfficeId, term) => {
                    officeIdUsedToSearch = actualOfficeId
                    termUsedToSearch = term
                    return availableBooksPromise
                },
                borrowBook: (bookId, name, email) => {
                    borrowedBookId = bookId
                    borrowerName = name
                    borrowerEmail = email
                }
            })

            typeOnSearchInput('a book', component)
            await availableBooksPromise
            clickOnFirstSearchResult(component.update())
            typeOnNameInput('adria', component)
            typeOnEmailInput('adria@pivotal.io', component)
            clickOnBorrowButton(component)
            done()
        })

        it('borrows book', () => {
            expect(officeIdUsedToSearch).toEqual(OFFICE_ID)
            expect(termUsedToSearch).toEqual('a book')
            expect(borrowedBookId).toEqual(1)
            expect(borrowerName).toEqual('adria')
            expect(borrowerEmail).toEqual('adria@pivotal.io')
        })
    })

    it('sets back location to office root', done => {
        mount(<BorrowBook
            office={{id: OFFICE_ID, name: OFFICE_NAME}}
            setBackLocation={location => {
                expect(location).toEqual(`/${OFFICE_NAME}`)
                done()
            }}
        />)
    })

    it('sets the header buttons as not visible', done => {
        mount(<BorrowBook
            office={{id: OFFICE_ID, name: OFFICE_NAME}}
            setHeaderConfig={config => {
                expect(config.displayButtons).toBeFalsy()
                done()
            }}
        />)
    })
})