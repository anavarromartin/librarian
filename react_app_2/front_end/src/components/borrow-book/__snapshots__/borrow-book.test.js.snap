// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BorrowBook /> renders according to snapshot 1`] = `
<BorrowBook
  borrowBook={[Function]}
  getAvailableBooks={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  office={
    Object {
      "id": 1,
      "name": "dallas",
    }
  }
  setBackLocation={[Function]}
  setHeaderVisibility={[Function]}
>
  <div
    className="borrow__top-container"
  >
    <h1
      className="borrow__title"
    >
      Borrow A Book
    </h1>
    <div
      className="borrow__find"
    >
      <label
        className="borrow__find-label"
      >
        Find an Available Book
      </label>
      <SearchInput
        isSearching={[Function]}
        onSelectResult={[Function]}
        search={[Function]}
      >
        <div
          className=""
        >
          <div
            className="search__field"
          >
            <div
              className="search__back"
              onClick={[Function]}
            >
              <FontAwesomeIcon
                border={false}
                className="search__back-icon"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [],
                      "f053",
                      "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
                    ],
                    "iconName": "chevron-left",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-chevron-left fa-w-10 search__back-icon"
                  data-icon="chevron-left"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </div>
            <div
              className="search__input"
            >
              <LabelInput
                clearLabel={[Function]}
                labels={Array []}
                onChange={[Function]}
                onClick={[Function]}
                value=""
              >
                <div
                  className="label-input__container"
                >
                  <div
                    className="labels__container"
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                  />
                  <input
                    className="labels-input__input"
                    onChange={[Function]}
                    onClick={[Function]}
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                    type="text"
                    value=""
                  />
                </div>
              </LabelInput>
            </div>
          </div>
          <div
            className="search__results"
            style={
              Object {
                "display": "none",
              }
            }
          >
            <SearchResults
              onSelectResult={[Function]}
              results={Array []}
            >
              <div
                className="search-results__top-container"
              />
            </SearchResults>
          </div>
        </div>
      </SearchInput>
    </div>
    <div
      className="borrow__name"
    >
      <label>
        Your Full Name
        <span
          style={
            Object {
              "color": "#bf1e25",
            }
          }
        >
          *
        </span>
      </label>
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="borrow__email"
    >
      <label>
        Your Email
        <span
          style={
            Object {
              "color": "#bf1e25",
            }
          }
        >
          *
        </span>
      </label>
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
    <button
      className="teal-button borrow__button"
      onClick={[Function]}
    >
      BORROW
    </button>
  </div>
</BorrowBook>
`;
